File: scarl-ui/src/game/equipment.js
import { List, Record } from 'immutable'

const OFF_HAND_DIVISOR = 2

const Slot = Record({
    key: undefined,
    label: undefined,
})

const Group = Record({
    prop: undefined,
    label: undefined,
    slots: undefined,
    fillAll: undefined,
    getSlotStats: undefined,
})

export const MainHand = 'Equipment.MainHand'
export const OffHand = 'Equipment.OffHand'
export const RangedSlot = 'Equipment.RangedSlot'
export const LauncherSlot = 'Equipment.LauncherSlot'
export const HeadArmor = 'Equipment.HeadArmor'
export const BodyArmor = 'Equipment.BodyArmor'
export const HandArmor = 'Equipment.HandArmor'
export const FootArmor = 'Equipment.FootArmor'

export const slots = {
    [MainHand]: Slot({
        key: MainHand,
        label: 'Main hand',
    }),
    [OffHand]: Slot({
        key: OffHand,
        label: 'Off-hand',
    }),
    [RangedSlot]: Slot({
        key: RangedSlot,
        label: 'Ranged weapon',
    }),
    [LauncherSlot]: Slot({
        key: LauncherSlot,
        label: 'Launcher',
    }),
    [HeadArmor]: Slot({
        key: HeadArmor,
        label: 'Head',
    }),
    [BodyArmor]: Slot({
        key: BodyArmor,
        label: 'Body',
    }),
    [HandArmor]: Slot({
        key: HandArmor,
        label: 'Hands',
    }),
    [FootArmor]: Slot({
        key: FootArmor,
        label: 'Feet',
    }),
}

export const groups = {
    Armor: Group({
        prop: 'armor',
        label: 'Wearable',
        slots: item => List([item.armor.slot]),
        fillAll: () => true,
        getSlotStats: item => item[groups.Armor.prop].stats,
    }),
    Launcher: Group({
        prop: 'launcher',
        label: 'Launcher',
        slots:() => List([LauncherSlot]),
        fillAll: () => true,
        getSlotStats: item => item[groups.Launcher.prop].stats,
    }),
    Shield: Group({
        prop: 'shield',
        label: 'Shield',
        slots: () => List([OffHand]),
        fillAll: () => true,
        getSlotStats: item => item[groups.Shield.prop].stats,
    }),
    RangedWeapon: Group({
        prop: 'rangedWeapon',
        label: 'Ranged weapon',
        slots:() => List([RangedSlot]),
        fillAll: () => true,
        getSlotStats: item => item[groups.RangedWeapon.prop].stats,
    }),
    Weapon: Group({
        prop: 'weapon',
        label: 'Melee weapon',
        slots: () => List([MainHand, OffHand]),
        fillAll: item => !!item.weapon.twoHanded,
        getSlotStats: (item, slot) => {
            const stats = item[groups.Weapon.prop].stats

            return ! item.weapon.twoHanded && slot === OffHand ? getOffHandWeaponStats(stats) : stats
        }
    }),
}

const getOffHandWeaponStats = stats => {
    const melee = stats.melee
    const consumption = melee.consumption

    return {
        ...stats,
        melee: {
            attack: Math.floor(melee.attack / OFF_HAND_DIVISOR),
            damage: Math.floor(melee.damage / OFF_HAND_DIVISOR),
            consumption: {
                energy: consumption.energy / OFF_HAND_DIVISOR,
                materials: consumption.materials / OFF_HAND_DIVISOR,
            },
        },
    }
}

File: scarl/src/main/scala/io/github/fiifoo/scarl/core/item/Equipment.scala
package io.github.fiifoo.scarl.core.item

import io.github.fiifoo.scarl.core.creature.Stats
import io.github.fiifoo.scarl.core.entity.Item
import io.github.fiifoo.scarl.core.item.Equipment._
import io.github.fiifoo.scarl.core.kind.ItemKind

object Equipment {

  sealed trait Category extends ItemKind.Category {
    def extractEquipment(item: ItemKind): Option[Equipment]
  }

  sealed trait ArmorCategory extends Category {
    def extractEquipment(item: ItemKind): Option[Equipment] = item.armor
  }

  case object HeadArmorCategory extends ArmorCategory

  case object BodyArmorCategory extends ArmorCategory

  case object HandArmorCategory extends ArmorCategory

  case object FootArmorCategory extends ArmorCategory

  case object LauncherCategory extends Category {
    def extractEquipment(item: ItemKind): Option[Equipment] = item.launcher
  }

  case object RangedWeaponCategory extends Category {
    def extractEquipment(item: ItemKind): Option[Equipment] = item.rangedWeapon
  }

  case object ShieldCategory extends Category {
    def extractEquipment(item: ItemKind): Option[Equipment] = item.shield
  }

  case object WeaponCategory extends Category {
    def extractEquipment(item: ItemKind): Option[Equipment] = item.weapon
  }

  val categories: Set[Category] = Set(
    HeadArmorCategory,
    BodyArmorCategory,
    HandArmorCategory,
    FootArmorCategory,
    LauncherCategory,
    RangedWeaponCategory,
    ShieldCategory,
    WeaponCategory,
  )

  sealed trait Slot

  sealed trait HandSlot extends Slot

  sealed trait ArmorSlot extends Slot

  case object MainHand extends HandSlot

  case object OffHand extends HandSlot

  case object RangedSlot extends Slot

  case object LauncherSlot extends Slot

  case object HeadArmor extends ArmorSlot

  case object BodyArmor extends ArmorSlot

  case object HandArmor extends ArmorSlot

  case object FootArmor extends ArmorSlot

  def selectEquipment(slot: Slot, item: Item): Option[Equipment] = {
    val valid: PartialFunction[Equipment, Equipment] = {
      case e: Equipment if e.slots contains slot => e
    }

    // Order might matter eg. two handed weapon + shield
    List(
      item.weapon collect valid,
      item.rangedWeapon collect valid,
      item.launcher collect valid,
      item.shield collect valid,
      item.armor collect valid
    ).flatten.headOption
  }
}

sealed trait Equipment {
  val stats: Stats
  val slots: Set[Slot]
  val fillAll: Boolean
  val category: Category
}

case class Armor(stats: Stats, slot: ArmorSlot) extends Equipment {
  val slots: Set[Slot] = Set(slot)
  val fillAll: Boolean = true
  val category: Category = slot match {
    case HeadArmor => HeadArmorCategory
    case BodyArmor => BodyArmorCategory
    case HandArmor => HandArmorCategory
    case FootArmor => FootArmorCategory
  }
}

case class MissileLauncher(stats: Stats) extends Equipment {
  val slots: Set[Slot] = Set(LauncherSlot)
  val fillAll: Boolean = true
  val category: Category = LauncherCategory
}

case class RangedWeapon(stats: Stats) extends Equipment {
  val slots: Set[Slot] = Set(RangedSlot)
  val fillAll: Boolean = true
  val category: Category = RangedWeaponCategory
}

case class Shield(stats: Stats) extends Equipment {
  val slots: Set[Slot] = Set(OffHand)
  val fillAll: Boolean = true
  val category: Category = ShieldCategory
}

case class Weapon(stats: Stats, twoHanded: Boolean) extends Equipment {
  val slots: Set[Slot] = Set(MainHand, OffHand)
  val fillAll: Boolean = twoHanded
  val category: Category = WeaponCategory
}
